utilities/pulumi_state_splitter/pulumi_state_splitter/split.py

```diff
@@ -0,0 +1,177 @@
+
+import pydantic
+import yaml
+
+import pulumi_state_splitter.fs
+import pulumi_state_splitter.model
+import pulumi_state_splitter.state_file
+import pulumi_state_splitter.stored_state
+
+
+class StateDir(pulumi_state_splitter.stored_state.StoredState):
+    """Represents a split Pulumi stack state."""
+
+    @property
+    def path(self) -> pathlib.Path:
+        """Path of the state directory."""
+        return self.backend_dir / str(self.stack_name)
+
+    @property
+    def _state_path(self):
+        return self.path / "state.yaml"
+
+    @classmethod
+    def from_state_file(cls, state_file: pulumi_state_splitter.state_file.StateFile):
+        """Converts a Pulumi stack state file to a split state."""
+        return cls(**state_file.model_dump())
+
+    def to_state_file(self) -> pulumi_state_splitter.state_file.StateFile:
+        """Converts a split state to a Pulumi stack state file."""
+        return pulumi_state_splitter.state_file.StateFile(**self.model_dump())
+
+    def remove(self):
+        self._state_path.unlink()
+        if self.state.checkpoint.latest:
+            for resource in self.state.checkpoint.latest.resources:
+                path = self.path / self.resource_subpath(resource)
+                if resource.type == "pulumi:pulumi:Stack":
+                    (self.path / "outputs.yaml").unlink()
+                path.unlink()
+                pulumi_state_splitter.fs.rmdir_if_empty(path.parent)
+        for d in (self.path, self.path.parent):
+            pulumi_state_splitter.fs.rmdir_if_empty(d)
+
+    @classmethod
+    def find(
+        cls, backend_dir: pathlib.Path
+    ) -> Iterable[pulumi_state_splitter.stored_state.StackName]:
+        """Finds all directory states in the Pulumi backend directory."""
+        glob_states = cls._glob_all()._state_path
+        for state_path in backend_dir.glob(str(glob_states)):
+            stack_dir = state_path.parent
+            yield pulumi_state_splitter.stored_state.StackName(
+                project=stack_dir.parent.name,
+                stack=stack_dir.name,
+            )
+
+    def load(self):
+        """Loads the contents of the state directory."""
+        with self._state_path.open() as f:
+            data = yaml.load(f, yaml.Loader)
+        self.state = pulumi_state_splitter.model.State.model_validate(data)
+        if not self.state.checkpoint.latest:
+            return
+        resources = []
+        for dirpath, _, filenames in os.walk(self.path):
+            dirpath = pathlib.Path(dirpath)
+            if dirpath == self.path:
+                continue
+            for filename in filenames:
+                with (dirpath / filename).open() as f:
+                    data = yaml.load(f, yaml.Loader)
+                resource = pulumi_state_splitter.model.Resource.model_validate(data)
+                if resource.type == "pulumi:pulumi:Stack":
+                    with (self.path / "outputs.yaml").open() as f:
```

Comment:
I wanted to have the stack's outputs easily available at `project/stack/outputs.yaml`.
Resource outputs are not that interesting on their own and there is a multitude of them - we'd effectively double the amount of YAML files if we did that.

Reply:
I meant exactly that - why bother with special handling for stack outputs and not keep them inline with the stack resource yaml :) But current approach works for me too.

Reply:
Because it would be buried deeper, at `project-name/stack-name/pulumi-pulumi-Stack/stack-name.yaml` instead of `project-name/stack-name/outputs.yaml`.

Alternatively we could just skip the type in the stack resource YAML path and always write it to `project-name/stack-name/stack.yaml`. But it still seems to me that it would be nice to have a dedicated file with just stack outputs.

WDYT?

Reply:
What we have currently looks good to me and now I understand the rationale. Thanks!